### PR TITLE
Searchy search search

### PR DESCRIPTION
**Pull request by DFurnes**
# Changes
- Adds client-side search, and some big performance improvements to the "drawer" view.

Building the interface on the client-side using React, because it's rendering model makes the gallery, drawer, and filtering interface waaaay easier to implement without tricky bugs.

Here's a [video of it in action](http://cl.ly/bXfs). :movie_camera: 

For review: @DoSomething/front-end

### CODE REVIEW

**Comment by DFurnes** [comment]
Ready for merge! 

Tested locally & there are no notable regressions from the old implementation. I'd like to revisit how we're building the voting form at some point so we don't have to use `dangerouslySetInnerHTML`, but it's not a blocker. A prerequisite for that (and another eventual nice-to-have) would be pre-rendering the client-side UI on the server using something like [react-laravel](https://github.com/talyssonoc/react-laravel).

File: app/Http/Controllers/CandidatesController.php [hunk 1]
```diff
@@ -40,8 +40,9 @@ public function index(Request $request)
         }
 
         $categories = Category::with('candidates')->get();
+        $categories_json = $categories->toJson();
```
Comment (sbsmith86) [inline app/Http/Controllers/CandidatesController.php:43]: oooo didn't know you could do this. nice.

Comment (DFurnes) [inline app/Http/Controllers/CandidatesController.php:43]: :smile:

**Comment by sbsmith86** [comment]
Whoa that video was coool. Looks good to me. Awesome we are using React in a project.

File: resources/assets/js/components/CandidateDetailView.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    // then grab that pre-rendered markup here. This code
+    // snippet is escaped server-side.
+    let formMarkup = document.getElementById('form-template').innerHTML;
+
+    // Replace token values for this particular candidate
+    formMarkup = formMarkup.replace(/CANDIDATE_ID/g, this.props.candidate.id)
+                           .replace(/CANDIDATE_NAME/g, this.props.candidate.name)
+                           .replace(/CANDIDATE_LINK/g, this.props.candidate.url)
+                           .replace(/TWITTER_NAME/g, this.props.candidate.share_name);
+
```
Comment (angaither) [inline resources/assets/js/components/CandidateDetailView.js:22]: YAAAAISSS

**Pull request merged**